# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of each sentence from `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
import re

def parse_sentences(sentences, vocab_to_int):
    
    unknown_id = vocab_to_int['<UNK>']
    
    sentences_as_ids = []
    for sentence in sentences:
        sentences_as_ids.append([])
        for word in sentence.split(' '):
            if vocab_to_int.get(word) == None:
                sentences_as_ids[-1].append(unknown_id)
            else:
                sentences_as_ids[-1].append(vocab_to_int[word])
    return sentences_as_ids

def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    print(source_vocab_to_int)
    target_text = target_text.replace('.', '. <EOS>')
    source_sentences = re.split('\n', source_text)
    target_sentences = re.split('\n', target_text)
    
    source_words_as_ids = parse_sentences(source_sentences, source_vocab_to_int)
    target_words_as_ids = parse_sentences(target_sentences, target_vocab_to_int)

    return source_words_as_ids, target_words_as_ids

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

{'<PAD>': 0, '<EOS>': 1, '<UNK>': 2, '<GO>': 3, 'mild': 4, 'september': 5, 'california': 6, 'june': 7, 'it': 8, 'march': 9, 'states': 10, 'autumn': 11, 'new': 12, 'jersey': 13, 'sometimes': 14, '.': 15, 'november': 16, 'is': 17, 'quiet': 18, ',': 19, 'chilly': 20, 'during': 21, 'usually': 22, 'in': 23, 'the': 24, 'hot': 25, 'united': 26, 'cold': 27, 'april': 28, 'july': 29, 'freezing': 30, 'snowy': 31, 'and': 32}
Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

{'<PAD>': 0, '<EOS>': 1, '<UNK>': 2, '<GO>': 3, 'grape.': 4, 'your': 5, 'how': 6, 'peaches.': 7, 'rusty': 8, 'their': 9, 'fall': 10, 'peach.': 11, 'when': 12, 'she': 13, 'october': 14, 'wonderful': 15, 'least': 16, 'june': 17, 'disliked': 18, "aren't": 19, 'mangoes.': 20, 'dry': 21, 'my': 22, 'monkeys': 23, 'where': 24, 'do': 25, 'horse': 26, 'paris': 27, 'went': 28, 'sometimes': 29, '.': 30, 'car': 31, 'pear': 32, 'wants': 33, 'want': 34, 'snake': 35, 'football': 36, 'oranges.': 37, ',': 38, 'never': 39, 'chilly': 40, 'old': 41, 'during': 42, 'beautiful': 43, 'france': 44, 'feared': 45, 'wet': 46, 'we': 47, 'dislike': 48, 'chinese': 49, 'shark': 50, 'pears.': 51, 'pleasant': 52, "isn't": 53, 'but': 54, 'liked': 55, 'spring': 56, 'likes': 57, 'liked.': 58, 'apple': 59, 'bird': 60, 'rabbit': 61, 'rainy': 62, 'plans': 63, 'pear.': 64, 'bananas.': 65, 'this': 66, 'english': 67, 'fruit.': 68, 'translate': 69, 'november': 70, 'limes.': 71, 'was': 72, 'grapefruit.': 73, 'orange': 74, 'mild':

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf

# Check TensorFlow Version
assert LooseVersion(tf.__version__) in [LooseVersion('1.0.0'), LooseVersion('1.0.1')], 'This project requires TensorFlow version 1.0  You are using {}'.format(tf.__version__)
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.0.0


/Users/mihkel/Documents/programming/software/anaconda3/envs/tensorflow/lib/python3.6/site-packages/ipykernel_launcher.py:14: UserWarning: No GPU found. Please use a GPU to train your neural network.
  


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoding_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.

Return the placeholders in the following the tuple (Input, Targets, Learing Rate, Keep Probability)

In [93]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, and learning rate.
    :return: Tuple (input, targets, learning rate, keep probability)
    """
    # TODO: Implement Function
    return (tf.placeholder(tf.int32, shape=[None, None], name='input'),
            tf.placeholder(tf.int32, shape=[None, None]),
            tf.placeholder(tf.float32),
            tf.placeholder(tf.float32, name='keep_prob'))

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoding Input
Implement `process_decoding_input` using TensorFlow to remove the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [104]:
def remove_last_element(target_data, batch_size):
    begin = tf.constant(0)
    shape = tf.TensorShape([target_data.shape[0], target_data.shape[1]-1])
    return tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])

def add_go(target_data, target_vocab_to_int, batch_size):
    return tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), target_data], 1)
    
def process_decoding_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for dencoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # helping example: https://github.com/udacity/deep-learning/blob/master/seq2seq/sequence_to_sequence_implementation.ipynb
    target_data = add_go(target_data, target_vocab_to_int, batch_size)
    target_data = remove_last_element(target_data, batch_size)
    return target_data

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_decoding_input(process_decoding_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer using [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn).

In [105]:
def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :return: RNN state
    """
    # TODO: Implement Function
    lstm = tf.contrib.rnn.BasicLSTMCell(rnn_size)
    drop = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
    cell = tf.contrib.rnn.MultiRNNCell([drop] * num_layers)
    _, state = tf.nn.dynamic_rnn(cell, rnn_inputs, dtype=tf.float32)
    return state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create training logits using [`tf.contrib.seq2seq.simple_decoder_fn_train()`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/simple_decoder_fn_train) and [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder).  Apply the `output_fn` to the [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder) outputs.

In [106]:
def decoding_layer_train(encoder_state, 
                         dec_cell, 
                         dec_embed_input, 
                         sequence_length, 
                         decoding_scope,
                         output_fn, 
                         keep_prob):
    from tensorflow.contrib.seq2seq import simple_decoder_fn_train, dynamic_rnn_decoder
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param sequence_length: Sequence Length
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_fn: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: Train Logits
    """

    # TODO: Implement Function
    dynamic_fn_train = simple_decoder_fn_train(encoder_state)
    outputs, final_state, final_context_state = dynamic_rnn_decoder(dec_cell, 
                                                  dynamic_fn_train, 
                                                  sequence_length=sequence_length,
                                                  scope=decoding_scope, 
                                                  inputs=dec_embed_input)
    return output_fn(outputs)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference logits using [`tf.contrib.seq2seq.simple_decoder_fn_inference()`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/simple_decoder_fn_inference) and [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder). 

In [107]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id,
                         maximum_length, vocab_size, decoding_scope, output_fn, keep_prob):
    from tensorflow.contrib.seq2seq import simple_decoder_fn_inference, dynamic_rnn_decoder
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param maximum_length: The maximum allowed time steps to decode
    :param vocab_size: Size of vocabulary
    :param decoding_scope: TensorFlow Variable Scope for decoding
    :param output_fn: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: Inference Logits
    """
    dynamic_fn_inference = simple_decoder_fn_inference(
                                output_fn=output_fn,
                                encoder_state=encoder_state, 
                                embeddings=dec_embeddings, 
                                start_of_sequence_id=start_of_sequence_id, 
                                end_of_sequence_id=end_of_sequence_id, 
                                maximum_length=maximum_length,
                                num_decoder_symbols=vocab_size)
    outputs, final_state, final_context_state = dynamic_rnn_decoder(dec_cell, dynamic_fn_inference)
    return outputs


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

- Create RNN cell for decoding using `rnn_size` and `num_layers`.
- Create the output fuction using [`lambda`](https://docs.python.org/3/tutorial/controlflow.html#lambda-expressions) to transform it's input, logits, to class logits.
- Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, sequence_length, decoding_scope, output_fn, keep_prob)` function to get the training logits.
- Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, maximum_length, vocab_size, decoding_scope, output_fn, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [108]:
def decoding_layer(dec_embed_input, dec_embeddings, encoder_state, vocab_size, sequence_length, rnn_size,
                   num_layers, target_vocab_to_int, keep_prob):
    """
    Create decoding layer
    :param dec_embed_input: Decoder embedded input
    :param dec_embeddings: Decoder embeddings
    :param encoder_state: The encoded state
    :param vocab_size: Size of vocabulary
    :param sequence_length: Sequence Length
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param keep_prob: Dropout keep probability
    :return: Tuple of (Training Logits, Inference Logits)
    """
    # TODO: Implement Function
    lstm = tf.contrib.rnn.BasicLSTMCell(rnn_size)
    drop = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
    dec_cell = tf.contrib.rnn.MultiRNNCell([drop] * num_layers)
    with tf.variable_scope("decoding") as decoding_scope:
        output_fn = lambda input_: tf.contrib.layers.fully_connected(input_, vocab_size, None, scope=decoding_scope)
        start_of_sequence_id = target_vocab_to_int['<GO>']
        end_of_sequence_id = target_vocab_to_int['<EOS>']
        maximum_length = 30
        training_logits = decoding_layer_train(encoder_state, dec_cell, dec_embed_input, sequence_length, decoding_scope, output_fn, keep_prob)
        inference_logits = decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, maximum_length, vocab_size, decoding_scope, output_fn, keep_prob)
    
    return training_logits, inference_logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Apply embedding to the input data for the encoder.
- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob)`.
- Process target data using your `process_decoding_input(target_data, target_vocab_to_int, batch_size)` function.
- Apply embedding to the target data for the decoder.
- Decode the encoded input using your `decoding_layer(dec_embed_input, dec_embeddings, encoder_state, vocab_size, sequence_length, rnn_size, num_layers, target_vocab_to_int, keep_prob)`.

In [109]:
def seq2seq_model(input_data, 
                  target_data, 
                  keep_prob, 
                  batch_size, 
                  sequence_length, 
                  source_vocab_size, 
                  target_vocab_size,
                  enc_embedding_size, 
                  dec_embedding_size, 
                  rnn_size, 
                  num_layers, 
                  target_vocab_to_int):
    from tensorflow.contrib.layers import embed_sequence 
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param sequence_length: Sequence Length
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training Logits, Inference Logits)
    """
    # TODO: Implement Function
    enc_embed_input = embed_sequence(input_data, source_vocab_size, enc_embedding_size)
    encoder_state = encoding_layer(enc_embed_input, rnn_size, num_layers, keep_prob)
    target_data = process_decoding_input(target_data, target_vocab_to_int, batch_size)
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, dec_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, target_data)
    return decoding_layer(dec_embed_input, 
                   dec_embeddings, 
                   encoder_state, 
                   target_vocab_size,
                   sequence_length, 
                   rnn_size, 
                   num_layers, 
                   target_vocab_to_int, 
                   keep_prob)



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability

In [110]:
# Number of Epochs
epochs = 1
# Batch Size
batch_size = 64
# RNN Size
rnn_size = 1
# Number of Layers
num_layers = 1
# Embedding Size
encoding_embedding_size = 64
decoding_embedding_size = 64
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 1

### Build the Graph
Build the graph using the neural network you implemented.

In [111]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_source_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob = model_inputs()
    sequence_length = tf.placeholder_with_default(max_source_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)
    
    train_logits, inference_logits = seq2seq_model(
        tf.reverse(input_data, [-1]), targets, keep_prob, batch_size, sequence_length, len(source_vocab_to_int), len(target_vocab_to_int),
        encoding_embedding_size, decoding_embedding_size, rnn_size, num_layers, target_vocab_to_int)

    tf.identity(inference_logits, 'logits')
    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            train_logits,
            targets,
            tf.ones([input_shape[0], sequence_length]))

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)

### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [112]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import time

def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1]), (0,0)],
            'constant')

    return np.mean(np.equal(target, np.argmax(logits, 2)))

train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]

valid_source = helper.pad_sentence_batch(source_int_text[:batch_size])
valid_target = helper.pad_sentence_batch(target_int_text[:batch_size])

with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch) in enumerate(
                helper.batch_data(train_source, train_target, batch_size)):
            start_time = time.time()
            
            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 sequence_length: target_batch.shape[1],
                 keep_prob: keep_probability})
            
            batch_train_logits = sess.run(
                inference_logits,
                {input_data: source_batch, keep_prob: 1.0})
            batch_valid_logits = sess.run(
                inference_logits,
                {input_data: valid_source, keep_prob: 1.0})
                
            train_acc = get_accuracy(target_batch, batch_train_logits)
            valid_acc = get_accuracy(np.array(valid_target), batch_valid_logits)
            end_time = time.time()
            print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.3f}, Validation Accuracy: {:>6.3f}, Loss: {:>6.3f}'
                  .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    0/2154 - Train Accuracy:  0.002, Validation Accuracy:  0.006, Loss:  5.871
Epoch   0 Batch    1/2154 - Train Accuracy:  0.017, Validation Accuracy:  0.015, Loss:  5.847
Epoch   0 Batch    2/2154 - Train Accuracy:  0.014, Validation Accuracy:  0.015, Loss:  5.820
Epoch   0 Batch    3/2154 - Train Accuracy:  0.012, Validation Accuracy:  0.015, Loss:  5.798
Epoch   0 Batch    4/2154 - Train Accuracy:  0.014, Validation Accuracy:  0.017, Loss:  5.768
Epoch   0 Batch    5/2154 - Train Accuracy:  0.014, Validation Accuracy:  0.018, Loss:  5.739
Epoch   0 Batch    6/2154 - Train Accuracy:  0.018, Validation Accuracy:  0.019, Loss:  5.708
Epoch   0 Batch    7/2154 - Train Accuracy:  0.017, Validation Accuracy:  0.019, Loss:  5.672
Epoch   0 Batch    8/2154 - Train Accuracy:  0.018, Validation Accuracy:  0.017, Loss:  5.652
Epoch   0 Batch    9/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.023, Loss:  5.614
Epoch   0 Batch   10/2154 - Train Accuracy:  0.035, Validati

Epoch   0 Batch   88/2154 - Train Accuracy:  0.225, Validation Accuracy:  0.221, Loss:  4.226
Epoch   0 Batch   89/2154 - Train Accuracy:  0.230, Validation Accuracy:  0.229, Loss:  4.243
Epoch   0 Batch   90/2154 - Train Accuracy:  0.227, Validation Accuracy:  0.229, Loss:  4.250
Epoch   0 Batch   91/2154 - Train Accuracy:  0.225, Validation Accuracy:  0.231, Loss:  4.187
Epoch   0 Batch   92/2154 - Train Accuracy:  0.231, Validation Accuracy:  0.228, Loss:  4.187
Epoch   0 Batch   93/2154 - Train Accuracy:  0.269, Validation Accuracy:  0.261, Loss:  4.211
Epoch   0 Batch   94/2154 - Train Accuracy:  0.267, Validation Accuracy:  0.259, Loss:  4.184
Epoch   0 Batch   95/2154 - Train Accuracy:  0.266, Validation Accuracy:  0.258, Loss:  4.148
Epoch   0 Batch   96/2154 - Train Accuracy:  0.270, Validation Accuracy:  0.261, Loss:  4.176
Epoch   0 Batch   97/2154 - Train Accuracy:  0.266, Validation Accuracy:  0.262, Loss:  4.170
Epoch   0 Batch   98/2154 - Train Accuracy:  0.276, Validati

Epoch   0 Batch  176/2154 - Train Accuracy:  0.513, Validation Accuracy:  0.492, Loss:  3.681
Epoch   0 Batch  177/2154 - Train Accuracy:  0.523, Validation Accuracy:  0.492, Loss:  3.668
Epoch   0 Batch  178/2154 - Train Accuracy:  0.501, Validation Accuracy:  0.492, Loss:  3.693
Epoch   0 Batch  179/2154 - Train Accuracy:  0.517, Validation Accuracy:  0.492, Loss:  3.632
Epoch   0 Batch  180/2154 - Train Accuracy:  0.487, Validation Accuracy:  0.492, Loss:  3.708
Epoch   0 Batch  181/2154 - Train Accuracy:  0.500, Validation Accuracy:  0.492, Loss:  3.665
Epoch   0 Batch  182/2154 - Train Accuracy:  0.503, Validation Accuracy:  0.492, Loss:  3.696
Epoch   0 Batch  183/2154 - Train Accuracy:  0.523, Validation Accuracy:  0.492, Loss:  3.551
Epoch   0 Batch  184/2154 - Train Accuracy:  0.531, Validation Accuracy:  0.492, Loss:  3.679
Epoch   0 Batch  185/2154 - Train Accuracy:  0.505, Validation Accuracy:  0.492, Loss:  3.770
Epoch   0 Batch  186/2154 - Train Accuracy:  0.496, Validati

Epoch   0 Batch  264/2154 - Train Accuracy:  0.380, Validation Accuracy:  0.361, Loss:  3.428
Epoch   0 Batch  265/2154 - Train Accuracy:  0.378, Validation Accuracy:  0.362, Loss:  3.440
Epoch   0 Batch  266/2154 - Train Accuracy:  0.368, Validation Accuracy:  0.363, Loss:  3.506
Epoch   0 Batch  267/2154 - Train Accuracy:  0.372, Validation Accuracy:  0.355, Loss:  3.539
Epoch   0 Batch  268/2154 - Train Accuracy:  0.359, Validation Accuracy:  0.348, Loss:  3.500
Epoch   0 Batch  269/2154 - Train Accuracy:  0.355, Validation Accuracy:  0.350, Loss:  3.514
Epoch   0 Batch  270/2154 - Train Accuracy:  0.361, Validation Accuracy:  0.351, Loss:  3.356
Epoch   0 Batch  271/2154 - Train Accuracy:  0.353, Validation Accuracy:  0.334, Loss:  3.463
Epoch   0 Batch  272/2154 - Train Accuracy:  0.351, Validation Accuracy:  0.333, Loss:  3.507
Epoch   0 Batch  273/2154 - Train Accuracy:  0.343, Validation Accuracy:  0.332, Loss:  3.399
Epoch   0 Batch  274/2154 - Train Accuracy:  0.345, Validati

Epoch   0 Batch  352/2154 - Train Accuracy:  0.382, Validation Accuracy:  0.365, Loss:  3.372
Epoch   0 Batch  353/2154 - Train Accuracy:  0.370, Validation Accuracy:  0.367, Loss:  3.344
Epoch   0 Batch  354/2154 - Train Accuracy:  0.378, Validation Accuracy:  0.367, Loss:  3.294
Epoch   0 Batch  355/2154 - Train Accuracy:  0.390, Validation Accuracy:  0.367, Loss:  3.423
Epoch   0 Batch  356/2154 - Train Accuracy:  0.386, Validation Accuracy:  0.366, Loss:  3.410
Epoch   0 Batch  357/2154 - Train Accuracy:  0.392, Validation Accuracy:  0.367, Loss:  3.238
Epoch   0 Batch  358/2154 - Train Accuracy:  0.384, Validation Accuracy:  0.367, Loss:  3.303
Epoch   0 Batch  359/2154 - Train Accuracy:  0.397, Validation Accuracy:  0.367, Loss:  3.359
Epoch   0 Batch  360/2154 - Train Accuracy:  0.382, Validation Accuracy:  0.367, Loss:  3.372
Epoch   0 Batch  361/2154 - Train Accuracy:  0.370, Validation Accuracy:  0.367, Loss:  3.457
Epoch   0 Batch  362/2154 - Train Accuracy:  0.388, Validati

Epoch   0 Batch  440/2154 - Train Accuracy:  0.256, Validation Accuracy:  0.250, Loss:  3.346
Epoch   0 Batch  441/2154 - Train Accuracy:  0.206, Validation Accuracy:  0.205, Loss:  3.339
Epoch   0 Batch  442/2154 - Train Accuracy:  0.199, Validation Accuracy:  0.192, Loss:  3.309
Epoch   0 Batch  443/2154 - Train Accuracy:  0.198, Validation Accuracy:  0.187, Loss:  3.262
Epoch   0 Batch  444/2154 - Train Accuracy:  0.176, Validation Accuracy:  0.169, Loss:  3.307
Epoch   0 Batch  445/2154 - Train Accuracy:  0.133, Validation Accuracy:  0.146, Loss:  3.409
Epoch   0 Batch  446/2154 - Train Accuracy:  0.142, Validation Accuracy:  0.135, Loss:  3.309
Epoch   0 Batch  447/2154 - Train Accuracy:  0.096, Validation Accuracy:  0.093, Loss:  3.329
Epoch   0 Batch  448/2154 - Train Accuracy:  0.080, Validation Accuracy:  0.077, Loss:  3.077
Epoch   0 Batch  449/2154 - Train Accuracy:  0.055, Validation Accuracy:  0.054, Loss:  3.270
Epoch   0 Batch  450/2154 - Train Accuracy:  0.012, Validati

Epoch   0 Batch  529/2154 - Train Accuracy:  0.189, Validation Accuracy:  0.177, Loss:  3.139
Epoch   0 Batch  530/2154 - Train Accuracy:  0.162, Validation Accuracy:  0.158, Loss:  3.228
Epoch   0 Batch  531/2154 - Train Accuracy:  0.167, Validation Accuracy:  0.162, Loss:  3.215
Epoch   0 Batch  532/2154 - Train Accuracy:  0.175, Validation Accuracy:  0.157, Loss:  3.078
Epoch   0 Batch  533/2154 - Train Accuracy:  0.161, Validation Accuracy:  0.155, Loss:  3.103
Epoch   0 Batch  534/2154 - Train Accuracy:  0.164, Validation Accuracy:  0.154, Loss:  3.186
Epoch   0 Batch  535/2154 - Train Accuracy:  0.153, Validation Accuracy:  0.153, Loss:  3.046
Epoch   0 Batch  536/2154 - Train Accuracy:  0.159, Validation Accuracy:  0.151, Loss:  3.190
Epoch   0 Batch  537/2154 - Train Accuracy:  0.155, Validation Accuracy:  0.147, Loss:  3.126
Epoch   0 Batch  538/2154 - Train Accuracy:  0.146, Validation Accuracy:  0.147, Loss:  3.260
Epoch   0 Batch  539/2154 - Train Accuracy:  0.153, Validati

Epoch   0 Batch  618/2154 - Train Accuracy:  0.116, Validation Accuracy:  0.106, Loss:  3.337
Epoch   0 Batch  619/2154 - Train Accuracy:  0.116, Validation Accuracy:  0.107, Loss:  3.235
Epoch   0 Batch  620/2154 - Train Accuracy:  0.114, Validation Accuracy:  0.107, Loss:  3.081
Epoch   0 Batch  621/2154 - Train Accuracy:  0.113, Validation Accuracy:  0.104, Loss:  3.184
Epoch   0 Batch  622/2154 - Train Accuracy:  0.114, Validation Accuracy:  0.105, Loss:  3.182
Epoch   0 Batch  623/2154 - Train Accuracy:  0.116, Validation Accuracy:  0.105, Loss:  3.217
Epoch   0 Batch  624/2154 - Train Accuracy:  0.116, Validation Accuracy:  0.104, Loss:  2.983
Epoch   0 Batch  625/2154 - Train Accuracy:  0.114, Validation Accuracy:  0.104, Loss:  3.220
Epoch   0 Batch  626/2154 - Train Accuracy:  0.117, Validation Accuracy:  0.104, Loss:  3.163
Epoch   0 Batch  627/2154 - Train Accuracy:  0.109, Validation Accuracy:  0.104, Loss:  3.170
Epoch   0 Batch  628/2154 - Train Accuracy:  0.116, Validati

Epoch   0 Batch  707/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.204
Epoch   0 Batch  708/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  3.243
Epoch   0 Batch  709/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.132
Epoch   0 Batch  710/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  3.098
Epoch   0 Batch  711/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.024, Loss:  3.110
Epoch   0 Batch  712/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.074
Epoch   0 Batch  713/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  3.132
Epoch   0 Batch  714/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.082
Epoch   0 Batch  715/2154 - Train Accuracy:  0.016, Validation Accuracy:  0.024, Loss:  3.014
Epoch   0 Batch  716/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.024, Loss:  2.974
Epoch   0 Batch  717/2154 - Train Accuracy:  0.023, Validati

Epoch   0 Batch  795/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  2.902
Epoch   0 Batch  796/2154 - Train Accuracy:  0.027, Validation Accuracy:  0.025, Loss:  3.161
Epoch   0 Batch  797/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  3.188
Epoch   0 Batch  798/2154 - Train Accuracy:  0.017, Validation Accuracy:  0.025, Loss:  3.137
Epoch   0 Batch  799/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  3.216
Epoch   0 Batch  800/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  3.181
Epoch   0 Batch  801/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  3.121
Epoch   0 Batch  802/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.189
Epoch   0 Batch  803/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.027
Epoch   0 Batch  804/2154 - Train Accuracy:  0.026, Validation Accuracy:  0.025, Loss:  3.091
Epoch   0 Batch  805/2154 - Train Accuracy:  0.020, Validati

Epoch   0 Batch  883/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.156
Epoch   0 Batch  884/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.090
Epoch   0 Batch  885/2154 - Train Accuracy:  0.018, Validation Accuracy:  0.025, Loss:  2.977
Epoch   0 Batch  886/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  2.957
Epoch   0 Batch  887/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.935
Epoch   0 Batch  888/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  3.175
Epoch   0 Batch  889/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  3.156
Epoch   0 Batch  890/2154 - Train Accuracy:  0.018, Validation Accuracy:  0.025, Loss:  3.095
Epoch   0 Batch  891/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  3.168
Epoch   0 Batch  892/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  3.149
Epoch   0 Batch  893/2154 - Train Accuracy:  0.027, Validati

Epoch   0 Batch  972/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  3.024
Epoch   0 Batch  973/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  3.068
Epoch   0 Batch  974/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  3.178
Epoch   0 Batch  975/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  3.131
Epoch   0 Batch  976/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  3.110
Epoch   0 Batch  977/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.088
Epoch   0 Batch  978/2154 - Train Accuracy:  0.019, Validation Accuracy:  0.025, Loss:  3.122
Epoch   0 Batch  979/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  2.963
Epoch   0 Batch  980/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  2.999
Epoch   0 Batch  981/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  3.052
Epoch   0 Batch  982/2154 - Train Accuracy:  0.024, Validati

Epoch   0 Batch 1061/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  3.032
Epoch   0 Batch 1062/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  3.039
Epoch   0 Batch 1063/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  3.096
Epoch   0 Batch 1064/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  3.018
Epoch   0 Batch 1065/2154 - Train Accuracy:  0.026, Validation Accuracy:  0.025, Loss:  3.055
Epoch   0 Batch 1066/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  3.002
Epoch   0 Batch 1067/2154 - Train Accuracy:  0.025, Validation Accuracy:  0.025, Loss:  3.043
Epoch   0 Batch 1068/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  3.086
Epoch   0 Batch 1069/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.104
Epoch   0 Batch 1070/2154 - Train Accuracy:  0.025, Validation Accuracy:  0.025, Loss:  2.948
Epoch   0 Batch 1071/2154 - Train Accuracy:  0.023, Validati

Epoch   0 Batch 1150/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  3.064
Epoch   0 Batch 1151/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  2.851
Epoch   0 Batch 1152/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  2.943
Epoch   0 Batch 1153/2154 - Train Accuracy:  0.019, Validation Accuracy:  0.025, Loss:  3.016
Epoch   0 Batch 1154/2154 - Train Accuracy:  0.019, Validation Accuracy:  0.025, Loss:  3.122
Epoch   0 Batch 1155/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.120
Epoch   0 Batch 1156/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  3.090
Epoch   0 Batch 1157/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  3.024
Epoch   0 Batch 1158/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  3.182
Epoch   0 Batch 1159/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  2.993
Epoch   0 Batch 1160/2154 - Train Accuracy:  0.022, Validati

Epoch   0 Batch 1238/2154 - Train Accuracy:  0.026, Validation Accuracy:  0.025, Loss:  3.004
Epoch   0 Batch 1239/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.124
Epoch   0 Batch 1240/2154 - Train Accuracy:  0.029, Validation Accuracy:  0.025, Loss:  3.119
Epoch   0 Batch 1241/2154 - Train Accuracy:  0.019, Validation Accuracy:  0.025, Loss:  2.965
Epoch   0 Batch 1242/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  2.975
Epoch   0 Batch 1243/2154 - Train Accuracy:  0.018, Validation Accuracy:  0.025, Loss:  3.021
Epoch   0 Batch 1244/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.095
Epoch   0 Batch 1245/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.019
Epoch   0 Batch 1246/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  3.129
Epoch   0 Batch 1247/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  3.074
Epoch   0 Batch 1248/2154 - Train Accuracy:  0.025, Validati

Epoch   0 Batch 1326/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  3.112
Epoch   0 Batch 1327/2154 - Train Accuracy:  0.025, Validation Accuracy:  0.025, Loss:  3.033
Epoch   0 Batch 1328/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  3.020
Epoch   0 Batch 1329/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  2.937
Epoch   0 Batch 1330/2154 - Train Accuracy:  0.019, Validation Accuracy:  0.025, Loss:  2.895
Epoch   0 Batch 1331/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  3.007
Epoch   0 Batch 1332/2154 - Train Accuracy:  0.027, Validation Accuracy:  0.025, Loss:  2.974
Epoch   0 Batch 1333/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.049
Epoch   0 Batch 1334/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  3.015
Epoch   0 Batch 1335/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  3.029
Epoch   0 Batch 1336/2154 - Train Accuracy:  0.023, Validati

Epoch   0 Batch 1414/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.886
Epoch   0 Batch 1415/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.936
Epoch   0 Batch 1416/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  3.089
Epoch   0 Batch 1417/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  2.932
Epoch   0 Batch 1418/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  3.036
Epoch   0 Batch 1419/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  2.907
Epoch   0 Batch 1420/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  3.023
Epoch   0 Batch 1421/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  2.940
Epoch   0 Batch 1422/2154 - Train Accuracy:  0.025, Validation Accuracy:  0.025, Loss:  2.965
Epoch   0 Batch 1423/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  3.045
Epoch   0 Batch 1424/2154 - Train Accuracy:  0.021, Validati

Epoch   0 Batch 1502/2154 - Train Accuracy:  0.028, Validation Accuracy:  0.025, Loss:  2.858
Epoch   0 Batch 1503/2154 - Train Accuracy:  0.026, Validation Accuracy:  0.025, Loss:  2.868
Epoch   0 Batch 1504/2154 - Train Accuracy:  0.025, Validation Accuracy:  0.025, Loss:  2.861
Epoch   0 Batch 1505/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.977
Epoch   0 Batch 1506/2154 - Train Accuracy:  0.026, Validation Accuracy:  0.026, Loss:  2.743
Epoch   0 Batch 1507/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.026, Loss:  3.030
Epoch   0 Batch 1508/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.026, Loss:  2.877
Epoch   0 Batch 1509/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.026, Loss:  2.897
Epoch   0 Batch 1510/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.026, Loss:  2.993
Epoch   0 Batch 1511/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.026, Loss:  2.938
Epoch   0 Batch 1512/2154 - Train Accuracy:  0.024, Validati

Epoch   0 Batch 1590/2154 - Train Accuracy:  0.026, Validation Accuracy:  0.025, Loss:  2.946
Epoch   0 Batch 1591/2154 - Train Accuracy:  0.025, Validation Accuracy:  0.025, Loss:  2.987
Epoch   0 Batch 1592/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  2.914
Epoch   0 Batch 1593/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  3.017
Epoch   0 Batch 1594/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  2.960
Epoch   0 Batch 1595/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  2.868
Epoch   0 Batch 1596/2154 - Train Accuracy:  0.026, Validation Accuracy:  0.025, Loss:  2.874
Epoch   0 Batch 1597/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.981
Epoch   0 Batch 1598/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.915
Epoch   0 Batch 1599/2154 - Train Accuracy:  0.025, Validation Accuracy:  0.025, Loss:  2.945
Epoch   0 Batch 1600/2154 - Train Accuracy:  0.024, Validati

Epoch   0 Batch 1678/2154 - Train Accuracy:  0.019, Validation Accuracy:  0.025, Loss:  2.910
Epoch   0 Batch 1679/2154 - Train Accuracy:  0.018, Validation Accuracy:  0.025, Loss:  2.874
Epoch   0 Batch 1680/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.843
Epoch   0 Batch 1681/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.956
Epoch   0 Batch 1682/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.816
Epoch   0 Batch 1683/2154 - Train Accuracy:  0.019, Validation Accuracy:  0.025, Loss:  2.843
Epoch   0 Batch 1684/2154 - Train Accuracy:  0.025, Validation Accuracy:  0.025, Loss:  2.877
Epoch   0 Batch 1685/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  3.015
Epoch   0 Batch 1686/2154 - Train Accuracy:  0.019, Validation Accuracy:  0.025, Loss:  2.804
Epoch   0 Batch 1687/2154 - Train Accuracy:  0.019, Validation Accuracy:  0.025, Loss:  2.999
Epoch   0 Batch 1688/2154 - Train Accuracy:  0.021, Validati

Epoch   0 Batch 1766/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.848
Epoch   0 Batch 1767/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  3.093
Epoch   0 Batch 1768/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  2.993
Epoch   0 Batch 1769/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  2.836
Epoch   0 Batch 1770/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  2.944
Epoch   0 Batch 1771/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  2.633
Epoch   0 Batch 1772/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  2.858
Epoch   0 Batch 1773/2154 - Train Accuracy:  0.027, Validation Accuracy:  0.025, Loss:  2.924
Epoch   0 Batch 1774/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  2.891
Epoch   0 Batch 1775/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  2.878
Epoch   0 Batch 1776/2154 - Train Accuracy:  0.022, Validati

Epoch   0 Batch 1854/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.858
Epoch   0 Batch 1855/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  2.835
Epoch   0 Batch 1856/2154 - Train Accuracy:  0.025, Validation Accuracy:  0.025, Loss:  2.933
Epoch   0 Batch 1857/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.984
Epoch   0 Batch 1858/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  2.726
Epoch   0 Batch 1859/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  2.871
Epoch   0 Batch 1860/2154 - Train Accuracy:  0.028, Validation Accuracy:  0.025, Loss:  3.002
Epoch   0 Batch 1861/2154 - Train Accuracy:  0.025, Validation Accuracy:  0.025, Loss:  2.921
Epoch   0 Batch 1862/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  2.801
Epoch   0 Batch 1863/2154 - Train Accuracy:  0.019, Validation Accuracy:  0.025, Loss:  2.769
Epoch   0 Batch 1864/2154 - Train Accuracy:  0.027, Validati

Epoch   0 Batch 1942/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  2.823
Epoch   0 Batch 1943/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  2.749
Epoch   0 Batch 1944/2154 - Train Accuracy:  0.020, Validation Accuracy:  0.025, Loss:  2.842
Epoch   0 Batch 1945/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.787
Epoch   0 Batch 1946/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.891
Epoch   0 Batch 1947/2154 - Train Accuracy:  0.019, Validation Accuracy:  0.025, Loss:  2.760
Epoch   0 Batch 1948/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.956
Epoch   0 Batch 1949/2154 - Train Accuracy:  0.029, Validation Accuracy:  0.025, Loss:  2.954
Epoch   0 Batch 1950/2154 - Train Accuracy:  0.025, Validation Accuracy:  0.025, Loss:  2.875
Epoch   0 Batch 1951/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.697
Epoch   0 Batch 1952/2154 - Train Accuracy:  0.024, Validati

Epoch   0 Batch 2030/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  2.872
Epoch   0 Batch 2031/2154 - Train Accuracy:  0.025, Validation Accuracy:  0.025, Loss:  2.873
Epoch   0 Batch 2032/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  3.013
Epoch   0 Batch 2033/2154 - Train Accuracy:  0.025, Validation Accuracy:  0.025, Loss:  2.807
Epoch   0 Batch 2034/2154 - Train Accuracy:  0.029, Validation Accuracy:  0.025, Loss:  2.888
Epoch   0 Batch 2035/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.823
Epoch   0 Batch 2036/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  2.953
Epoch   0 Batch 2037/2154 - Train Accuracy:  0.025, Validation Accuracy:  0.025, Loss:  2.661
Epoch   0 Batch 2038/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  2.767
Epoch   0 Batch 2039/2154 - Train Accuracy:  0.026, Validation Accuracy:  0.025, Loss:  2.938
Epoch   0 Batch 2040/2154 - Train Accuracy:  0.021, Validati

Epoch   0 Batch 2118/2154 - Train Accuracy:  0.023, Validation Accuracy:  0.025, Loss:  2.998
Epoch   0 Batch 2119/2154 - Train Accuracy:  0.026, Validation Accuracy:  0.025, Loss:  2.993
Epoch   0 Batch 2120/2154 - Train Accuracy:  0.017, Validation Accuracy:  0.025, Loss:  2.905
Epoch   0 Batch 2121/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  2.834
Epoch   0 Batch 2122/2154 - Train Accuracy:  0.018, Validation Accuracy:  0.025, Loss:  2.678
Epoch   0 Batch 2123/2154 - Train Accuracy:  0.019, Validation Accuracy:  0.025, Loss:  2.790
Epoch   0 Batch 2124/2154 - Train Accuracy:  0.021, Validation Accuracy:  0.025, Loss:  2.851
Epoch   0 Batch 2125/2154 - Train Accuracy:  0.025, Validation Accuracy:  0.025, Loss:  2.855
Epoch   0 Batch 2126/2154 - Train Accuracy:  0.024, Validation Accuracy:  0.025, Loss:  2.776
Epoch   0 Batch 2127/2154 - Train Accuracy:  0.022, Validation Accuracy:  0.025, Loss:  2.782
Epoch   0 Batch 2128/2154 - Train Accuracy:  0.024, Validati

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [ ]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [ ]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [ ]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    return None


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

## Translate
This will translate `translate_sentence` from English to French.

In [ ]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('logits:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence], keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in np.argmax(translate_logits, 1)]))
print('  French Words: {}'.format([target_int_to_vocab[i] for i in np.argmax(translate_logits, 1)]))

## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.